src/arrow-util/src/builder.rs

```diff
@@ -129,72 +128,130 @@ impl ArrowBuilder {
             if length.map_or(false, |l| l.into_u32() < i32::MAX.unsigned_abs()) {
-                DataType::Utf8
+                (DataType::Utf8, "char")
             } else {
-                DataType::LargeUtf8
+                (DataType::LargeUtf8, "char")
             }
         }
         ScalarType::VarChar { max_length } => {
             if max_length.map_or(false, |l| l.into_u32() < i32::MAX.unsigned_abs()) {
-                DataType::Utf8
+                (DataType::Utf8, "varchar")
             } else {
-                DataType::LargeUtf8
+                (DataType::LargeUtf8, "varchar")
             }
         }
-        ScalarType::String => DataType::LargeUtf8,
-        ScalarType::MzTimestamp => {
-            // MZ timestamps are in milliseconds
-            DataType::Timestamp(arrow::datatypes::TimeUnit::Millisecond, None)
+        ScalarType::String => (DataType::LargeUtf8, "string"),
+        // Parquet does have a UUID 'Logical Type' in parquet format 2.4+, but there is no arrow
+        // UUID type, so we match the format (a 16-byte fixed-length binary array) ourselves.
+        ScalarType::Uuid => (DataType::FixedSizeBinary(16), "uuid"),
+        // Parquet does have a JSON 'Logical Type' in parquet format 2.4+, but there is no arrow
+        // JSON type, so for now we represent JSON as 'large' utf8-encoded strings.
+        ScalarType::Jsonb => (DataType::LargeUtf8, "jsonb"),
+        ScalarType::MzTimestamp => (DataType::UInt64, "mztimestamp"),
+        ScalarType::Numeric { max_scale } => {
+            // Materialize allows 39 digits of precision for numeric values, but allows
+            // arbitrary scales among those values. e.g. 1e38 and 1e-39 are both valid in
+            // the same column. However, Arrow/Parquet only allows static declaration of both
+            // the precision and the scale. To represent the full range of values of a numeric
+            // column, we would need 78-digits to store all possible values. Arrow's Decimal256
+            // type can only support 76 digits, so we are be unable to represent the entire range.
+
+            // Instead of representing the full possible range, we instead try to represent most
+            // values in the most-compatible way. We use a Decimal128 type which can handle 38
+            // digits of precision and has more compatibility with other parquet readers than
+            // Decimal256. We use Arrow's default scale of 10 if max-scale is not set. We will
+            // error if we encounter a value that is too large to represent, and if that happens
+            // a user can choose to cast the column to a string to represent the value.
+            match max_scale {
+                Some(scale) => {
+                    let scale = i8::try_from(scale.into_u8()).expect("known <= 39");
+                    if scale <= 38 {
+                        (
+                            DataType::Decimal128(DECIMAL128_MAX_PRECISION, scale),
```

Comment:
I think we need to make sure `scale < DECIMAL128_MAX_PRECISION`, maybe?

Reply:
Ah there's actually an arrow `DECIMAL128_MAX_SCALE` I can compare against, which is also 38. (precision / scale can both be 38 when we want to represent a number where all digits are past the decimal point)